[icon-themes] Support glob keys in file icon associations

zm-cttae opened the pull request:
Closes #12493.

## Background

The feature request is for glob support in the key of icon mappings for file icon themes

This PR allows multiple use cases - see table:

| Key to target a file or folder | Example filename |
| --- | --- |
| `*` | `Makefile`, `Dockerfile` |
| `Jenkinsfile.*` | `Jenkinsfile.prod` |
| `test_*.py` | `test_component.py` |
| `*_test.go` | `component_test.go` |
| `*Test.cs` | `CalculatorTest.cs` |
| `.env.*` | `.env.development` |
| `web.*.config` | `web.prod.config` |
| `webpack.*.config.js` | `webpack.dev.config.js` |
| `*.webpack.config.js` | `shared.dev.webpack.config.js` |
| `.env.*.*` | `.env.development.docker.example.env` |
| `*.md` | `pop goes ...................................... the weasel.md` |

## Specification

File icon theme keys are separated by the last dot into `basename` and `ext`.

The `basename` segment allows any number of `.` characters.

Both segments can be globbed like so:

*   Wildcard glob `*` accepts any value.
*   Prefix glob `prefix*`.
*   Suffix glob `*suffix`.
*   Prefix-suffix glob `*middle*`.
*   Middle glob `prefix*suffix`

## Implementation

File/folder icon association keys that can be augmented by globs: `folderNames`, `folderNamesExpanded`, `fileNames`

Any `*` in the key of the file/folder icon mapping:

*   generates a different CSS attribute selector for `basename` and/or `extname` attr values in the file icon theme
*   removes the +2 score boost contributed by the name class

This PR introduces **extraAttributes** option to icon label APIs. These attributes are anchors that the CSS from the file icon theme side can target. This is necessary to prevent a [FOUC](https://en.wikipedia.org/wiki/Flash_of_unstyled_content) after opening any collapsed folder or any text document, caused by file icon themes running 100s of globs per each file label.

EDIT:

> This was updated to remove `**` from this spec and sync with `vscode.GlobPattern` spec.  
> It previously meant >=2 dot segments - https://github.com/microsoft/vscode/pull/174286#issuecomment-1462494545

EDIT 2:

> Apply changeset from https://github.com/microsoft/vscode/pull/174286#issuecomment-1464506764

EDIT 3:

> Redesign implementation in line with #177650 - moves business logic of globs back to the frontend

zm-cttae committed:
Support globs in file icon themes (#12493)

zm-cttae commented:
@microsoft-github-policy-service agree

zm-cttae committed:
Fix CI compile error

zm-cttae committed:
Merge branch 'main' into feat-file-icon-theme-glob

aeschli commented:
Can you explain what css classes you generate?

zm-cttae committed:
Fix post-glob ext file icon classes (#174286)

zm-cttae commented:
Done in PR description

zm-cttae commented:
Done in PR description

zm-cttae commented:
While we are here:

*   `-name-dir-icon` was not the proposed CSS class suffix in #136656 and the current `-name-dir-icon` looks ugly..
*   Can I make a new PR to clean that up (into `-dir-${kind}-icon`)?

aeschli commented:
Sorry, can you give examples how this works (maybe a table covering the various cases, prefix, suffix, `**`)
- file/folder pattern in the icon theme
- all classes in the css rules that are generated

Give some example of resources in the explorer and how the class names match. Show that file name matches are always stronger than file extension matches and stronger than fold name matches.

zm-cttae committed:
Fix coaelesced glob regex

zm-cttae commented:
Fair enough: [updated replit with table](https://replit.com/@zm-cttae/microsoft-vscode-12493#index.js)

<table>
<caption></caption>
<tr><th>File name</th><th>Glob classes (<code>getIconClasses</code>)</th><th>Time (ms)</th></tr>
<tr><td><code>"Makefile"</code></td><td><code>*-glob-file-icon</code><td>0.3</td></tr>
<tr><td><code>"Makefile"</code></td><td><code>*-glob-file-icon</code><td>0.0</td></tr>
<tr><td><code>"Jenkinsfile.prod"</code></td><td><code>**-glob-file-icon</code><br/><code>*.*-glob-file-icon</code><br/><code>Jenkinsfile.*-glob-file-icon</code><br/><code>*.prod-glob-file-icon</code><td>0.1</td></tr>
<tr><td><code>"test_component.py"</code></td><td><code>**-glob-file-icon</code><br/><code>*.*-glob-file-icon</code><br/><code>test_component.*-glob-file-icon</code><br/><code>*.py-glob-file-icon</code><br/><code>test_*.py-glob-file-icon</code><br/><code>*_component.py-glob-file-icon</code><td>0.1</td></tr>
<tr><td><code>"component_test.go"</code></td><td><code>**-glob-file-icon</code><br/><code>*.*-glob-file-icon</code><br/><code>component_test.*-glob-file-icon</code><br/><code>*.go-glob-file-icon</code><br/><code>component_*.go-glob-file-icon</code><br/><code>*_test.go-glob-file-icon</code><td>0.0</td></tr>
<tr><td><code>".env.development"</code></td><td><code>**-glob-file-icon</code><br/><code>*.*.*-glob-file-icon</code><br/><code>.**-glob-file-icon</code><br/><code>.*.*-glob-file-icon</code><br/><code>*.env.*-glob-file-icon</code><br/><code>.env.*-glob-file-icon</code><br/><code>**.development-glob-file-icon</code><br/><code>*.*.development-glob-file-icon</code><br/><code>.*.development-glob-file-icon</code><br/><code>*.env.development-glob-file-icon</code><td>0.0</td></tr>
<tr><td><code>"web.prod.config"</code></td><td><code>**-glob-file-icon</code><br/><code>*.*.*-glob-file-icon</code><br/><code>web.**-glob-file-icon</code><br/><code>web.*.*-glob-file-icon</code><br/><code>*.prod.*-glob-file-icon</code><br/><code>web.prod.*-glob-file-icon</code><br/><code>**.config-glob-file-icon</code><br/><code>*.*.config-glob-file-icon</code><br/><code>web.*.config-glob-file-icon</code><br/><code>*.prod.config-glob-file-icon</code><td>0.0</td></tr>
<tr><td><code>"webpack.dev.config.js"</code></td><td><code>**-glob-file-icon</code><br/><code>*.*.*.*-glob-file-icon</code><br/><code>webpack.**-glob-file-icon</code><br/><code>webpack.*.*.*-glob-file-icon</code><br/><code>*.dev.**-glob-file-icon</code><br/><code>*.dev.*.*-glob-file-icon</code><br/><code>webpack.dev.**-glob-file-icon</code><br/><code>webpack.dev.*.*-glob-file-icon</code><br/><code>**.config.*-glob-file-icon</code><br/><code>*.*.config.*-glob-file-icon</code><br/><code>webpack.*.config.*-glob-file-icon</code><br/><code>*.dev.config.*-glob-file-icon</code><br/><code>webpack.dev.config.*-glob-file-icon</code><br/><code>**.js-glob-file-icon</code><br/><code>*.*.*.js-glob-file-icon</code><br/><code>webpack.**.js-glob-file-icon</code><br/><code>webpack.*.*.js-glob-file-icon</code><br/><code>*.dev.*.js-glob-file-icon</code><br/><code>webpack.dev.*.js-glob-file-icon</code><br/><code>**.config.js-glob-file-icon</code><br/><code>*.*.config.js-glob-file-icon</code><br/><code>webpack.*.config.js-glob-file-icon</code><br/><code>*.dev.config.js-glob-file-icon</code><td>0.0</td></tr>
<tr><td><code>"shared.dev.webpack.config.js"</code></td><td><code>shared.**-glob-file-icon</code><br/><code>*.dev.webpack.config.js-glob-file-icon</code><br/><code>shared.dev.**-glob-file-icon</code><br/><code>shared.*.webpack.config.js-glob-file-icon</code><br/><code>shared.dev.webpack.**-glob-file-icon</code><br/><code>shared.dev.*.config.js-glob-file-icon</code><br/><code>shared.dev.webpack.config.*-glob-file-icon</code><br/><code>shared.dev.webpack.*.js-glob-file-icon</code><td>0.0</td></tr>
<tr><td><code>".env.development.docker.example.env"</code></td><td><code>.**-glob-file-icon</code><br/><code>*.env.development.docker.example.env-glob-file-icon</code><br/><code>.env.**-glob-file-icon</code><br/><code>.*.development.docker.example.env-glob-file-icon</code><br/><code>.env.development.**-glob-file-icon</code><br/><code>.env.*.docker.example.env-glob-file-icon</code><br/><code>.env.development.docker.**-glob-file-icon</code><br/><code>.env.development.*.example.env-glob-file-icon</code><br/><code>.env.development.docker.example.*-glob-file-icon</code><br/><code>.env.development.docker.*.env-glob-file-icon</code><td>0.0</td></tr>
<tr><td><code>"pop goes ...................................... the weasel.md"</code></td><td><code>**-glob-file-icon</code><br/><code>*.*-glob-file-icon</code><br/><code>pop goes  the weasel.*-glob-file-icon</code><br/><code>*.md-glob-file-icon</code><td>0.0</td></tr>
</table>

zm-cttae commented:
<table>
<caption><h3>File icon theme input in <code>fileIconThemeData</code> service</h3></caption>
<tr><th>Key to target a file</th><th>Example filename</th></tr>
<tr><td><code>*</code></td><td><code>Makefile</code></td></tr>
<tr><td><code>Jenkinsfile.*</code></td><td><code>Jenkinsfile.prod</code></td></tr>
<tr><td><code>*_component.py</code></td><td><code>test_component.py</code></td></tr>
<tr><td><code>component_*.py</code></td><td><code>component_test.go</code></td></tr>
<tr><td><code>.env.*</code></td><td><code>.env.development</code></td></tr>
<tr><td><code>web.*.config</code></td><td><code>web.prod.config</code></td></tr>
<tr><td><code>webpack.*.config.js</code></td><td><code>webpack.dev.config.js</code></td></tr>
<tr><td><code>**.webpack.config.js</code></td><td><code>shared.dev.webpack.config.js</code></td></tr>
<tr><td><code>.env.**</code></td><td><code>.env.development.docker.example.env</code></td></tr>
<tr><td><code>*.md</code></td><td><code>pop goes ...................................... the weasel.md</code></td></tr>
</table>

zm-cttae committed:
Narrow folder name type in getIconClasses

zm-cttae committed:
Polishing

zm-cttae committed:
Update CSS glob comments post-polish

zm-cttae committed:
Typo: file -> folder

zm-cttae committed:
Polish up variable names

zm-cttae commented:
Made some small tweaks in my work lunch, but please feel free to review now @aeschli

zm-cttae committed:
let -> const

zm-cttae committed:
Merge branch 'main' into feat-file-icon-theme-glob

aeschli commented:
ok, now I understand. So `*` is only possible at the beginning or end of dot, underscore or minus separated segments.

The problem is to make sure the matching order holds:

'full file match' beats 'file match with pattern' beats 'file extension match' beats 'file extension with pattern' beats language beats folder name beats folder name with

How this is implemented is the number of classes matching

IMO we're at a dead end with CSS class name matching. I'm reluctant in adding even more class names.

zm-cttae commented:
I think this is possible in `fileIconThemeData` side.. but our loop/condtional depth would reach 6... 👀

- We would build two arrays in the for loops inc. `folderNames`, `folderNamesExpanded`, `fileNames` icon mapping keys.
- Globs would go first - these all contain a '*' character in theme key
- Identifiers would go next - these don't contain a '*' character in theme key

zm-cttae commented:
I've had a change of tack ("let's use `name-${kind}-icon`") and we can bypass this issue in a saner manner

zm-cttae committed:
Reduce glob icon priority of using name score classes
Refs: [#issuecomment-1429874059](github.com/microsoft/vscode/pull/174286#issuecomment-1429874059)

zm-cttae committed:
Cleanup pass on CSS glob comments

zm-cttae committed:
Fix regression (folder-specifc icons mixing with files)

aeschli commented:
How does this PR help https://github.com/vscode-icons/vscode-icons/issues/2754?

M00TSDNS commented:
https://github.com/microsoft/vscode/blob/7d0f02a/src/vs/editor/common/services/getIconClasses.ts#L130-L164

zm-cttae commented:
https://github.com/microsoft/vscode/blob/7d0f02a34d1f83e4b800bbcd9efcfcb8ab5b0afd/src/vs/editor/common/services/getIconClasses.ts#L130-L164

zm-cttae committed:
Remove switch for dash separator

zm-cttae committed:
let -> const

zm-cttae committed:
Cleanup pass

zm-cttae commented:
This looks ready to ship :shipit:

zm-cttae committed:
Apply "1 file extension" restriction to dash/underscore globs

zm-cttae committed:
[nit] use icon kind in dirname per #136656
Make it clear where the dirname class is contributed from.

JimiC commented:
If this gets included in `vscode` it will be a major milestone for the icon theme support and a heaven sent gift for all icon theme extension authors.

We have been waiting for this like forever :smile_cat:

zm-cttae commented:
@aeschli can I drop the `**` for multiple dots and just use `*` instead?  
When used in `vscode.GlobPattern` API, the `**` coalescing wildcard means "multiple slashes".

If we use `*` to mean "any number of dots", it's consistent to the glob API in the rest of the app.

aeschli commented:
I'm all in favour of simplifying the syntax (and capabilities). The suggested solution is not trivial, complicates the current implementation but still doesn't provide the full glob support that (I suspect) file-icon-theme authors are hoping for.

One idea is to only support `test_*.py` (basename-prefix-with-undescore) and `*_test.go (basename-suffix-with-undescore). 
But then users will ask about support for `-`. camelCase, other separators...

If you have any ideas, let's discuss before implementing.

zm-cttae committed:
Use `*` (non-coalescing wildcard) to match multiple dots
This commit brings consistency with `vscode.GlobPattern` API.
Refs: #174286 @aeschli review

zm-cttae commented:
> But then users will ask about support for `-`. camelCase, other separators...  

Because of laravel+php I will be implementing `CamelCaseTest` boundary separator. That one seems trivial...

aeschli commented:
If you want this PR to be merged, we first need to first communicate and agree on an approach. 
I want a proposal that is easy to understand for users and has a simple implementation that we can maintain.

I recommend you to wait with implementing before we agree.

zm-cttae commented:
Okay. Would you like the commits on CamelCase and regexp class move to be reverted?
Happy to do so if you wish, it's your Owned Product!

> One idea is to only support test_*.py (basename-prefix-with-undescore) and *_test.go (basename-suffix-with-undescore).

I feel the `.` separator is the most important use case - if you'd like to limit the PR to just that, we can impl+test `.` separator only.

zm-cttae commented:
You might still think "well, `.` is very much covered by filename extensions, so lets just add dash support."  
If so, I'm happy to go ahead and just limit ourselves to `basename-*-with-underscore` & `non-coalescing-wildcard` support.

zm-cttae committed:
Remove permutative and wildcard chain globs
This limits the spec to:
- `non-coalescing-wildcard`
- `basename-prefix-with-undescore`
- `basename-suffix-with-undescore`
Refs: #174286

zm-cttae committed:
Ban icon globs for >=5 filename dot segments
Refs: #174286

zm-cttae commented:
@aeschli please look at d06ed11 where I aggressively cut down the spec to a sane level (3 independent glob features).

*   `non-coalescing-wildcard`
*   `basename-prefix-with-undescore`
*   `basename-suffix-with-undescore`

## Simplified specification

- For a filename with max 4 dot segments, can use `*` to match any 1 dot segment
- ~~`*` works as an alias for two or more `*` dot segments in an icon key (coaelescedGlob)~~
- ~~More than 4 dot segments reduces output significantly to $O(N)$ (2n-2) linear implementation:~~
  - ~~Match any one segment with a `*` wildcard (baseGlob)~~
  - ~~Match any amount of dot segments via prefix - uses `*` wildcard (suffixGlob)~~
- Support for globbing dash separator with `segment.ext` using prefix and suffix globs (e.g. `test_*.py` or `*_test.go`)
  - Suffix glob using `test_*.py` or `test-*.py`
  - Prefix glob using `*-test.go` or `*-test.go`

### Supported input in `fileIconThemeData` service

| Key to target a file or folder | Example filename |   |
| --- | --- | --- |
| `*` | `Makefile` |   |
| `Jenkinsfile.*` | `Jenkinsfile.prod` |   |
| `test_*.py` (special case for `segment.ext`) | `test_component.py` |   |
| `*_test.go` (special case for `segment.ext`) | `component_test.go` |   |
| `.env.*` | `.env.development` |   |
| `web.*.config` | `web.prod.config` |   |
| `webpack.*.config.js` | `webpack.dev.config.js` |   |
| `*.md` | `pop goes ...................................... the weasel.md` |   |

- Removal: `.env.*` ->`.env.development.docker.example.env`
- Removal: glob support for >=5 filename dot segments

zm-cttae committed:
Note limit in filename extension comments

zm-cttae committed:
Limit icon glob wildcard to file extensions (>=2nd segment)
Refs: #174286

zm-cttae committed:
Fix dash separated segment globs

zm-cttae committed:
[nit] Move basenameGlob comment above #segments!=2 if-return

zm-cttae committed:
[nit] Test typeof filename is string for icon classes

zm-cttae committed:
Merge branch 'main' into feat-file-icon-theme-glob

zm-cttae committed:
[nit] Rename base-glob vars to wildcard-glob

zm-cttae committed:
Merge branch 'main' into feat-file-icon-theme-glob

aeschli commented:
@zm-cttae Sorry I didn't get back earlier, it was quite a busy week. 

I like your last change. Thanks a lot!

I wanted to bring up an idea. Could we use the [CSS Substring matching selectors](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Selectors/Attribute_selectors#substring_matching_selectors)) to further reduce the number of classes needed and allow some more ways on how theme authors can use the `*`?

See https://github.com/microsoft/vscode/issues/177650 to see a sketch of this.

What do you think?

zm-cttae committed:
Merge pull request #1 from zm-cttae/pr-174286/hotfix-issuecomment-1463879783

Apply feedback from PR microsoft#174286

seanCodes commented:
Hi @zm-cttae, is this something you're still working on?

zm-cttae commented:
This is ready to go but we have a better impl that @aeschli demoed - not clear whether that's been packaged up. We're also both busy, it seems. (Currently there's endgame for VS Code & I'm wrangling with Textmate library migration).

zm-cttae committed:
Merge branch 'microsoft:main' into feat-file-icon-theme-glob

thelooter commented:
Whats the status on this?

Repiteo commented:
Hoping this gets off the backburner soon. A lack of glob patterns has been the most obvious weakpoint of the icon API for nearly a decade, and it feels like we're *so close* to finally fixing that.

muchisx commented:
🙏🏼 Looking forward for this being implemented soon @zm-cttae @aeschli

JoaquinFernandez commented:
Hope to see this merged soon!

zm-cttae committed:
WIP - implement simpler file icon glob assocations (#177650)

zm-cttae committed:
Merge branch 'main' into feat-file-icon-theme-glob

zm-cttae committed:
Fix missing import

zm-cttae committed:
Fix dirname refactor

zm-cttae committed:
Wire up icon attributes for file icon globs

zm-cttae committed:
Fix readonly type error in icon label options

zm-cttae committed:
Wire up chat widget labels

zm-cttae committed:
Merge branch 'main' into feat-file-icon-theme-glob

zm-cttae committed:
Merge branch 'main' into feat-file-icon-theme-glob

zm-cttae committed:
Fix pushGlobSelectors compile error

zm-cttae committed:
Escape handleParentFolder CSS string

zm-cttae committed:
Get CI passing

zm-cttae committed:
Merge branch 'main' into feat-file-icon-theme-glob

zm-cttae committed:
Merge branch 'main' into feat-file-icon-theme-glob

zm-cttae committed:
[nit] Fix indent in chatInlineAnchorWidget.ts

zm-cttae committed:
Merge branch 'main' into feat-file-icon-theme-glob

adreo00 commented:
Looking forward to this PR merging!

zm-cttae committed:
Merge branch 'main' into feat-file-icon-theme-glob

zm-cttae committed:
Restore #241508

zm-cttae committed:
Improve CSS class escaping

zm-cttae committed:
Handle single glob key e.g. `folder/*`

zm-cttae committed:
Merge branch 'main' into feat-file-icon-theme-glob

zm-cttae committed:
Fix globs for fileNames key

aeschli commented:
@zm-cttae Thanks for all the work.

I didn't realize the ripples we get from adding `extraAttributes`. `
`getIconClasses` needs to return typed object, e.g. `FileIconInfo`, that we keep as long as possible, elgl in the 

I think its best to tackle that first, in a separate PR.

zm-cttae commented:
I take it the signaturr is, then will be `{ classes, [attributes ...] }`?

arkthur commented:
Hi! Thanks for all the work done so far.
That said, is there anything I can help with? What is stalling this?

willi84 commented:
cannot wait for that feature. btw. is this capable to separat between workspaces?

ViniDevBR commented:
hey. and about this?
Could resolve the conflicts? to continue CI
